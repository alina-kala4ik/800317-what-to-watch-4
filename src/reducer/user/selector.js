import {NameSpace} from "./../name-space.js";

const getAuthorizationStatus = (state) => {
  return state[NameSpace.USER].authorizationStatus;
};

const getAvatar = (state) => {
  return state[NameSpace.USER].avatar;
};

const getFlagIsFetchingAuthStatus = (state) => {
  return state[NameSpace.USER].isFetchingAuthStatus;
};

const getFlagLoginDataValid = (state) => {
  return state[NameSpace.USER].isLoginDataValid;
}

export {getAuthorizationStatus, getAvatar, getFlagIsFetchingAuthStatus, getFlagLoginDataValid};
