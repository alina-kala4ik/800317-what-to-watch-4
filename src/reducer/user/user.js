import {extend} from "./../../utils.js";
import {ActionCreator as AppStateActionCreator} from "./../app-state/app-state.js";

const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

const initialState = {
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  avatar: null,
};


const ActionTypes = {
  REQUIRED_AUTHORIZATION: `REQUIRED_AUTHORIZATION`,
  ADD_AVATAR: `ADD_AVATAR`,
};

const ActionCreator = {
  requiredAuthorization: (status)=>({
    type: ActionTypes.REQUIRED_AUTHORIZATION,
    payload: status,
  }),
  addAvatar: (url)=>({
    type: ActionTypes.ADD_AVATAR,
    payload: url
  }),
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionTypes.REQUIRED_AUTHORIZATION:
      return extend(state, {
        authorizationStatus: action.payload
      });
    case ActionTypes.ADD_AVATAR:
      return extend(state, {
        avatar: action.payload
      });
  }
  return state;
};

const Operation = {
  checkAuth: () => (dispatch, getState, api) => {
    return api.get(`/login`)
      .then(()=>{
        dispatch(ActionCreator.requiredAuthorization(AuthorizationStatus.AUTH));
      })
      .catch((err)=>{
        throw err;
      });
  },

  login: (authData) => (dispatch, getState, api) => {
    return api.post(`/login`, {
      email: authData.email,
      password: authData.password
    })
      .then((response)=>{
        const {avatar_url: avatarUrl} = response.data;

        dispatch(ActionCreator.requiredAuthorization(AuthorizationStatus.AUTH));
        dispatch(ActionCreator.addAvatar(avatarUrl));
        dispatch(AppStateActionCreator.logIn(false));
      })
      .catch((err)=>{
        throw err;
      });
  }
};

export {ActionCreator, reducer, Operation, AuthorizationStatus, ActionTypes};
