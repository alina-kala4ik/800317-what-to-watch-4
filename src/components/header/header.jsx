import React from "react";
import {Link} from "react-router-dom";
import {Pages} from "./../../utils.js";
import {connect} from "react-redux";
import {getAuthorizationStatus, getAvatar} from "./../../reducer/user/selector.js";
import {AuthorizationStatus} from "./../../reducer/user/user.js";
import PropTypes from "prop-types";

const Header = (props) => {
  const {authorizationStatus, avatar, children, uniqueClasses = ``, isActiveLogoLink} = props;

  const logo = <React.Fragment>
    <span className="logo__letter logo__letter--1">W</span>
    <span className="logo__letter logo__letter--2">T</span>
    <span className="logo__letter logo__letter--3">W</span>
  </React.Fragment>;

  const logoLink = isActiveLogoLink ?
    <Link
      className="logo__link"
      to={Pages.ROOT}
    >
      {logo}
    </Link> :
    <a className="logo__link">
      {logo}
    </a>;

  const userBlock = authorizationStatus === AuthorizationStatus.AUTH ?
    <Link
      className="user-block__avatar"
      to={Pages.MY_LIST}>
      <img src={avatar} alt="User avatar" width="63" height="63"/>
    </Link> :
    <Link
      to={Pages.LOGIN}
      className="user-block__link">
      Sign in
    </Link>;

  return <header className={`page-header ${uniqueClasses}`}>
    <div className="logo">
      {logoLink}
    </div>

    {children}

    <div className="user-block">
      {userBlock}
    </div>
  </header>;
};

const mapStateToProps = (state) => ({
  authorizationStatus: getAuthorizationStatus(state),
  avatar: getAvatar(state),
});

Header.propTypes = {
  authorizationStatus: PropTypes.string.isRequired,
  avatar: PropTypes.string,
  children: PropTypes.node,
  uniqueClasses: PropTypes.string,
  isActiveLogoLink: PropTypes.bool.isRequired,
};

export {Header};
export default connect(mapStateToProps)(Header);

