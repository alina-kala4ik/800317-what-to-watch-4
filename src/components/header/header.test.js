import React from "react";
import renderer from "react-test-renderer";
import {Header} from "./header.jsx";
import {AuthorizationStatus} from "./../../reducer/user/user.js";
import {Router} from "react-router-dom";
import history from "./../../history.js";

it(`render Header when user authorized`, ()=>{
  const tree = renderer.create(
      <Router history={history}>
        <Header
          authorizationStatus={AuthorizationStatus.AUTH}
          avatar={``}
          isActiveLogoLink={false}
        />
      </Router>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it(`render Header when user unauthorized`, ()=>{
  const tree = renderer.create(
      <Router history={history}>
        <Header
          authorizationStatus={AuthorizationStatus.NO_AUTH}
          avatar={``}
          isActiveLogoLink={false}
        />
      </Router>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it(`render Header with children`, ()=>{
  const tree = renderer.create(
      <Router history={history}>
        <Header
          authorizationStatus={AuthorizationStatus.NO_AUTH}
          avatar={``}
          isActiveLogoLink={false}
        >
          <div></div>
        </Header>
      </Router>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it(`render Header with active logo and unique classes from header`, ()=>{
  const tree = renderer.create(
      <Router history={history}>
        <Header
          authorizationStatus={AuthorizationStatus.NO_AUTH}
          avatar={``}
          isActiveLogoLink={true}
          uniqueClasses={`uniqueClasses`}
        >
          <div></div>
        </Header>
      </Router>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});


