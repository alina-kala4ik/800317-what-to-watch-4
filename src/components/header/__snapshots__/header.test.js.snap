// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Header when user authorized 1`] = `
<header
  className="page-header "
>
  <div
    className="logo"
  >
    <a
      className="logo__link"
    >
      <span
        className="logo__letter logo__letter--1"
      >
        W
      </span>
      <span
        className="logo__letter logo__letter--2"
      >
        T
      </span>
      <span
        className="logo__letter logo__letter--3"
      >
        W
      </span>
    </a>
  </div>
  <div
    className="user-block"
  >
    <a
      className="user-block__avatar"
      href="/mylist"
      onClick={[Function]}
    >
      <img
        alt="User avatar"
        height="63"
        src=""
        width="63"
      />
    </a>
  </div>
</header>
`;

exports[`render Header when user unauthorized 1`] = `
<header
  className="page-header "
>
  <div
    className="logo"
  >
    <a
      className="logo__link"
    >
      <span
        className="logo__letter logo__letter--1"
      >
        W
      </span>
      <span
        className="logo__letter logo__letter--2"
      >
        T
      </span>
      <span
        className="logo__letter logo__letter--3"
      >
        W
      </span>
    </a>
  </div>
  <div
    className="user-block"
  >
    <a
      className="user-block__link"
      href="/login"
      onClick={[Function]}
    >
      Sign in
    </a>
  </div>
</header>
`;

exports[`render Header with active logo and unique classes from header 1`] = `
<header
  className="page-header uniqueClasses"
>
  <div
    className="logo"
  >
    <a
      className="logo__link"
      href="/"
      onClick={[Function]}
    >
      <span
        className="logo__letter logo__letter--1"
      >
        W
      </span>
      <span
        className="logo__letter logo__letter--2"
      >
        T
      </span>
      <span
        className="logo__letter logo__letter--3"
      >
        W
      </span>
    </a>
  </div>
  <div />
  <div
    className="user-block"
  >
    <a
      className="user-block__link"
      href="/login"
      onClick={[Function]}
    >
      Sign in
    </a>
  </div>
</header>
`;

exports[`render Header with children 1`] = `
<header
  className="page-header "
>
  <div
    className="logo"
  >
    <a
      className="logo__link"
    >
      <span
        className="logo__letter logo__letter--1"
      >
        W
      </span>
      <span
        className="logo__letter logo__letter--2"
      >
        T
      </span>
      <span
        className="logo__letter logo__letter--3"
      >
        W
      </span>
    </a>
  </div>
  <div />
  <div
    className="user-block"
  >
    <a
      className="user-block__link"
      href="/login"
      onClick={[Function]}
    >
      Sign in
    </a>
  </div>
</header>
`;

exports[`render Header without user block 1`] = `
<header
  className="page-header uniqueClasses"
>
  <div
    className="logo"
  >
    <a
      className="logo__link"
      href="/"
      onClick={[Function]}
    >
      <span
        className="logo__letter logo__letter--1"
      >
        W
      </span>
      <span
        className="logo__letter logo__letter--2"
      >
        T
      </span>
      <span
        className="logo__letter logo__letter--3"
      >
        W
      </span>
    </a>
  </div>
  <div />
</header>
`;
